server/stats.go

```diff
@@ -0,0 +1,228 @@
+		time.Sleep(dither)
+
+		ticker := time.NewTicker(statsAutosaveInterval)
+		for {
+			select {
+			case _ = <-stop:
+				return
+			case _ = <-ticker.C:
+				p.saveStats()
+			}
+		}
+	}()
+
+	p.updateConfig(func(c *config) {
+		c.statsStopAutosave = stop
+	})
+}
+
+// To save the stats periodically, use `go Autosave(...)``
+func (p *Plugin) stopAutosaveStats(conf config) {
+	stop := conf.statsStopAutosave
+	if stop == nil {
+		return
+	}
+	p.updateConfig(func(c *config) {
+		c.statsStopAutosave = nil
+	})
+	stop <- true
+}
+
+func (p *Plugin) saveStats() error {
+	stats := p.getConfig().stats
+	if stats == nil {
+		return nil
+	}
+	data, err := json.Marshal(stats)
+	if err != nil {
+		return err
+	}
+	appErr := p.API.KVSetWithExpiry(statsKeyName(), data, int64(statsKeyExpiration.Seconds()))
+	if appErr != nil {
+		return appErr
+	}
+	return nil
+}
+
+// This is only useful in a single-server context, so can not be used in production
+// TODO: Need a way to reset all stats in production?
+func (p *Plugin) debugResetStats() error {
+	conf := p.getConfig()
+	p.stopAutosaveStats(conf)
+	defer p.startAutosaveStats()
+
+	stats := conf.stats
+	if stats == nil {
+		return nil
+	}
+	stats.Reset()
+
+	for i := 0; ; i++ {
+		keys, appErr := p.API.KVList(i, listPerPage)
+		if appErr != nil {
+			return appErr
+		}
+		for _, key := range keys {
+			if !strings.HasPrefix(key, prefixStats) {
+				continue
+			}
+			appErr := p.API.KVDelete(key)
+			if appErr != nil {
+				return appErr
+			}
+		}
+		if len(keys) < listPerPage {
+			break
+		}
+	}
+
+	return nil
+}
+
+func initUserCounter(currentInstanceStore CurrentInstanceStore, userStore UserStore) {
+	goexpvar.Publish("jira/mapped_users", goexpvar.Func(func() interface{} {
+		ji, err := currentInstanceStore.LoadCurrentJIRAInstance()
+		if err != nil {
+			return err.Error()
+		}
+		c, err := userStore.CountUsers(ji)
+		if err != nil {
+			return err.Error()
+		}
+		return c
+	}))
+}
+
+var startedAt = time.Now()
+
+// initUptime adds an "uptime" expvar
+func initUptime() {
+	goexpvar.Publish("uptime", goexpvar.Func(func() interface{} {
+		up := time.Since(startedAt).Round(time.Second)
+		return up.String()
+	}))
+}
+
+func (p *Plugin) consolidatedStoredStats() (*expvar.Stats, []string, error) {
+	stats := expvar.NewUnpublishedStats(nil)
+	var statsKeys []string
+	for i := 0; ; i++ {
+		keys, appErr := p.API.KVList(i, listPerPage)
+		if appErr != nil {
+			return nil, nil, appErr
+		}
+
+		for _, key := range keys {
+			if !strings.HasPrefix(key, prefixStats) {
+				continue
+			}
+			var data []byte
+			data, appErr = p.API.KVGet(key)
+			if appErr != nil {
+				return nil, nil, appErr
+			}
+			from := expvar.NewUnpublishedStats(data)
+			stats.Merge(from)
+			statsKeys = append(statsKeys, key)
+		}
+
+		if len(keys) < listPerPage {
+			break
+		}
+	}
+	return stats, statsKeys, nil
+}
+
+func statsKeyName() string {
+	hostname, _ := os.Hostname()
```

Comment:
What if the hostname changes?

Reply:
stats pull together all keys that match the pattern. Each key is set to expire in 1 month, so in an installation that rotates through hostnames regularly, you'd get 1 month worth of stats.

We need a better stats backend available to plugins that this. Thoughts?

Reply:
We could expose some kind of server supported method. Where you write your stats and the server takes care of it. It has access to unique ids for that purpose.

Reply:
See https://community-daily.mattermost.com/core/pl/8fwujudf7bnzxk1qgfzukjd9ww - my personal preference would be to expose statsd (or similar) running as a plugin, and listen on its own port. All that would be needed is exposing an IsLeader (or a `LeaderHasBeenElected` notification, along with a `p.API.IsLeader`).